### PR TITLE
[CCR] Add time since last auto follow fetch to auto follow stats

### PR DESCRIPTION
**Pull request by martijnvg**
For each remote cluster the auto follow coordinator, starts an auto
follower that checks the remote cluster state and determines whether an
index needs to be auto followed. The time since last auto follow is
reported per remote cluster and gives insight whether the auto follow
process is alive.

Relates to #33007
Originates from #35895

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ccr/AutoFollowStats.java [hunk 1]
```diff
@@ -73,24 +90,28 @@ public static AutoFollowStats fromXContent(final XContentParser parser) {
     private final long numberOfFailedRemoteClusterStateRequests;
     private final long numberOfSuccessfulFollowIndices;
     private final NavigableMap<String, ElasticsearchException> recentAutoFollowErrors;
+    private final NavigableMap<String, Long> trackingRemoteClusters;
```
Comment (martijnvg) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ccr/AutoFollowStats.java]: I wonder whether all stats should be reported per remote cluster, so `numberOfSuccessfulFollowIndices`, `recentAutoFollowErrors` etc. would then be reported on a per remote cluster bases. This makes more sense to me now considering how the auto follow coordinator currently operates.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java [hunk 2]
```diff
@@ -239,20 +259,25 @@ public void clusterChanged(ClusterChangedEvent event) {
         private final String remoteCluster;
         private final Consumer<List<AutoFollowResult>> statsUpdater;
         private final Supplier<ClusterState> followerClusterStateSupplier;
+        private final LongSupplier relativeTimeProvider;
 
+        private volatile long lastAutoFollowTime = -1;
```
Comment (dnhatn) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java]: Can we add the time unit to the variable names (relativeTimeProvider and lastAutoFollowTime)?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ccr/AutoFollowStats.java [hunk 3]
```diff
@@ -73,24 +90,28 @@ public static AutoFollowStats fromXContent(final XContentParser parser) {
...
     public AutoFollowStats(
-            long numberOfFailedFollowIndices,
-            long numberOfFailedRemoteClusterStateRequests,
-            long numberOfSuccessfulFollowIndices,
-            NavigableMap<String, ElasticsearchException> recentAutoFollowErrors
+        long numberOfFailedFollowIndices,
+        long numberOfFailedRemoteClusterStateRequests,
+        long numberOfSuccessfulFollowIndices,
+        NavigableMap<String, ElasticsearchException> recentAutoFollowErrors,
+        NavigableMap<String, Long> trackingRemoteClusters
```
Comment (dnhatn) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ccr/AutoFollowStats.java]: Not sure about this parameter name (`trackingRemoteClusters`).

**Review by dnhatn — COMMENTED** [review]
Thanks @martijnvg. This looks good. I left some comments.

**Comment by martijnvg** [comment]
@dnhatn Thanks for look and I've updated the PR.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java [hunk 4]
```diff
@@ -239,20 +264,25 @@ public void clusterChanged(ClusterChangedEvent event) {
         private final String remoteCluster;
         private final Consumer<List<AutoFollowResult>> statsUpdater;
         private final Supplier<ClusterState> followerClusterStateSupplier;
+        private final LongSupplier relativeTimeProvider;
 
+        private volatile long lastAutoFollowTimeInNanos = -1;
```
Comment (dnhatn) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java]: This also became ms now.

**Review by dnhatn — APPROVED** [review]
I left some comments around the time unit. Feel free to merge it after addressing these.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java [hunk 5]
```diff
@@ -79,10 +84,13 @@
     public AutoFollowCoordinator(
             Client client,
             ClusterService clusterService,
-            CcrLicenseChecker ccrLicenseChecker) {
+            CcrLicenseChecker ccrLicenseChecker,
+            LongSupplier relativeNanoTimeProvider) {
```
[the remark this answers is not in the document]
Comment (martijnvg) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java]: Yikes, I completely missed that. Good catch!

**Review by dnhatn — APPROVED** [review]
LGTM

**Comment by martijnvg** [comment]
run the gradle build tests 1

**Comment by bleskes** [comment]
I wonder if we should call this `time_since_last_check`. Fetch sounds implementation specific IMO.

**Comment by codebrain** [comment]
Would it be possible to update the documentation? https://www.elastic.co/guide/en/elasticsearch/reference/6.7/ccr-get-stats.html

**Pull request merged**